File: sdk/sdk-js/src/exceptions.ts
/**
 * Lattice SDK Exceptions
 *
 * Custom exceptions for the Lattice SDK.
 */

/**
 * Base exception for all Lattice SDK errors.
 */
export class LatticeError extends Error {
    public statusCode?: number;

    constructor(message: string, statusCode?: number) {
        super(message);
        this.name = "LatticeError";
        this.statusCode = statusCode;
    }
}

/**
 * Raised when unable to connect to the Lattice server.
 */
export class LatticeConnectionError extends LatticeError {
    public originalError?: Error;

    constructor(message: string, originalError?: Error) {
        super(message);
        this.name = "LatticeConnectionError";
        this.originalError = originalError;
    }
}

/**
 * Raised when the API returns an error response.
 */
export class LatticeApiError extends LatticeError {
    public errorMessage?: string;

    constructor(message: string, statusCode: number, errorMessage?: string) {
        super(message, statusCode);
        this.name = "LatticeApiError";
        this.errorMessage = errorMessage || message;
    }
}

/**
 * Raised when request validation fails.
 */
export class LatticeValidationError extends LatticeError {
    public field?: string;

    constructor(message: string, field?: string) {
        super(message);
        this.name = "LatticeValidationError";
        this.field = field;
    }
}

File: sdk/sdk-python/lattice_sdk/exceptions.py
"""
Lattice SDK Exceptions

Custom exceptions for the Lattice SDK.
"""

from typing import Optional


class LatticeException(Exception):
    """Base exception for all Lattice SDK errors."""

    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class LatticeConnectionError(LatticeException):
    """Raised when unable to connect to the Lattice server."""

    def __init__(self, message: str, original_error: Optional[Exception] = None):
        super().__init__(message)
        self.original_error = original_error


class LatticeApiError(LatticeException):
    """Raised when the API returns an error response.

    The new Lattice error contract is a raw JSON body of the shape
    ``{ "error": "<message>", "detail"?: <structured> }`` accompanied by a
    non-2xx HTTP status code. ``detail`` carries any structured error
    information (e.g. schema validation errors, lock metadata) and
    ``request_id`` is populated from the ``X-Lattice-Request-Id`` response
    header when present.
    """

    def __init__(
        self,
        message: str,
        status_code: int,
        error_message: Optional[str] = None,
        detail: Optional[object] = None,
        request_id: Optional[str] = None,
    ):
        super().__init__(message, status_code)
        self.error_message = error_message or message
        self.detail = detail
        self.request_id = request_id


class LatticeValidationError(LatticeException):
    """Raised when request validation fails."""

    def __init__(self, message: str, field: Optional[str] = None):
        super().__init__(message)
        self.field = field
